driver-core/src/main/com/mongodb/internal/TimeoutContext.java

```diff
@@ -136,11 +138,8 @@ public long timeoutOrAlternative(final long alternativeTimeoutMS) {
         Long timeoutMS = timeoutSettings.getTimeoutMS();
         if (timeoutMS == null) {
             return alternativeTimeoutMS;
-        } else if (timeoutMS == 0) {
-            return timeoutMS;
-        } else {
-            return timeoutRemainingMS();
         }
```

Comment:
`else if (timeoutMS == 0)` is removed due to a logic issue where using an original infinite timeout (timeoutMS = 0) incorrectly led to always returning 0 from TimeoutSettings in `timeoutOrAlternative`, regardless of actual `computedServerSelectionTimeout` time left. Now, `timeoutRemainingMS()` handles both infinite and finite timeouts, checking against the `timeout` object itself:

For the reference: 
`  private long timeoutRemainingMS() {
        assertNotNull(timeout);
        if (timeout.hasExpired()) {
            throw createMongoTimeoutException("The operation timeout has expired.");
        }
        return timeout.isInfinite() ? 0 : timeout.remaining(MILLISECONDS);
    }`